storage/google_cloud/google_cloud_utils.py

```diff
@@ -84,9 +86,31 @@ def upload_file_to_blob(bucket_credentials_file_path, target_blob_url, f):
     :type target_blob_url: str
     :param f: File to upload, opened in binary mode.
     :type f: file-like
+    :param max_retries: maximum number of times to retry uploading the file.
+    :type max_retries: int
+    :param blob_chunk_size: the size of a chunk of data whenever iterating (in MiB).
+    :type blob_chunk_size: float
     """
-    log.info(f"Uploading file to blob '{target_blob_url}'...")
-    storage_client = storage.Client.from_service_account_json(bucket_credentials_file_path)
-    blob = _blob_at_url(storage_client, target_blob_url)
-    blob.upload_from_file(f)
-    log.info(f"Uploaded file to blob")
+    try:
+        log.info(f"Uploading file to blob '{target_blob_url}'...")
+        storage_client = storage.Client.from_service_account_json(bucket_credentials_file_path)
+        blob = _blob_at_url(storage_client, target_blob_url)
+        blob.chunk_size = blob_chunk_size * 1024 * 1024
+        blob.upload_from_file(f)
+        log.info(f"Uploaded file to blob")
+        upload_status = "success"
+
+    except (ConnectionError, socket.timeout, Timeout):
+        log.warning("Failed to upload due to connection error!")
+        if max_retries > 0:
+            log.info(f"Retrying {max_retries} more times with a reduced chunk_size of {blob_chunk_size - 30}MiB")
```

Comment:
If you just subtract 30 you risk the chunk size going negative

Reply:
The 3 max no of retries helps us prevent this.